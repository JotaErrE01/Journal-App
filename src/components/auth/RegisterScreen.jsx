import React from 'react'
import { Link } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import validator from 'validator';
import { useDispatch } from 'react-redux';
import { removeError, setError } from '../../actions/ui';
import { useSelector } from 'react-redux';
import { starRegisterWithEmailPassworName } from '../../actions/auth';

const RegisterScreen = () => {

    const dispacth = useDispatch();
    const error = useSelector(state => state.ui.msgError);

    const [ usuario, handleInputChange ] = useForm({
        name: '',
        email: '',
        password: '',
        password2: ''
    });

    const { name, email, password, password2 } = usuario;

    const handleRegister = e => {
        e.preventDefault();

        if(isFormValid()){
            dispacth( removeError() );
            dispacth( starRegisterWithEmailPassworName( email, password, name ) );
        }
    }

    // validar si el formulario es valido
    const isFormValid = () => {

        if( name.trim() === '' ) {
            dispacth( setError( 'Name is required' ) );
            return false;
        }

        if( !validator.isEmail(email) ){
            dispacth( setError( 'Email is not valid' ) );
            return false;
        }

        if( password.length < 5 || password !== password2 ){
            dispacth( setError( 'Password should be at least 5 characters and match each others' ) );
            return false;
        }

        return true;
    }

    return (
        <div
            className="animate__animated animate__fadeIn animate__faster"
        >
            <h3 className='auth__title mb-5'>Register</h3>

            <form
                onSubmit={handleRegister}
            >

                {
                    error &&
                        <div className="auth__alert-error">
                            { error }
                        </div>
                }

                <input  
                    type="text" 
                    placeholder="Name"
                    name="name" 
                    className='auth__input mb-5'
                    autoComplete="off"
                    value={name}
                    onChange={handleInputChange}
                />

                <input  
                    type="text" 
                    placeholder="Email"
                    name="email" 
                    className='auth__input mb-5'
                    autoComplete="off"
                    value={email}
                    onChange={handleInputChange}
                />

                <input 
                    type="password" 
                    placeholder="Password"
                    name="password"
                    className='auth__input mb-5'
                    autoComplete="off"
                    value={password}
                    onChange={handleInputChange}
                />

                <input 
                    type="password" 
                    placeholder="Conifrm Password"
                    name="password2"
                    className='auth__input mb-5'
                    autoComplete="off"
                    value={password2}
                    onChange={handleInputChange}
                />

                <button
                    type="submit"
                    className='btn btn-primary pointer btn-block mb-5'
                >Register</button>

                <Link to='/auth/login' className="link" >Already Register</Link>
            </form>
        </div>
    )
}

export default RegisterScreen;