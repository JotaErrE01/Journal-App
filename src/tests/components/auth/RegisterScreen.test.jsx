import { render, screen } from "@testing-library/react";
import configureStore from 'redux-mock-store';
import { Provider } from "react-redux";
import RegisterScreen from "../../../components/auth/RegisterScreen";
import thunk from "redux-thunk";
import { MemoryRouter } from "react-router";
import userEvent from "@testing-library/user-event";

describe('Pruebas en el componente <LoginScreen />', () => {

    const mockStore = configureStore([ thunk ]);
    const initState = {
        auth: { },
        ui: {
            loading: false,
            msg: null
        }
    }

    let store;

    let wrapper;

    beforeEach( () => {

        store = mockStore( initState );
        
        wrapper = render(
            <MemoryRouter>
                <Provider store= {store}>
                    <RegisterScreen />
                </Provider>
            </MemoryRouter>
        );
    } );

    
    test('Debe de hacer match con el sanpshot', () => {
        
        expect( wrapper.container ).toMatchSnapshot();

    })

    test('Debe de hacer el dispatch de la accion respectiva', () => {
        
        const { container } = wrapper;
        const [ name, email, pass, pass2 ] = container.querySelectorAll( '.auth__input' );

        userEvent.type( name, '' );
        userEvent.type( email, 'correo@correo.com' );
        userEvent.type( pass, '123456' );
        userEvent.type( pass2, '123456' );
        
        userEvent.click( container.querySelector('.btn-block') );

        const actions = store.getActions();

        expect( actions.length ).toBe( 1 );

    })    

    test('Debe de mostrar el mensaje de error', () => {

        const initState = {
            auth: { },
            ui: {
                loading: false,
                msgError: 'Nombre es obligatorio'
            }
        }

        store = mockStore( initState );
        
        wrapper = render(
            <MemoryRouter>
                <Provider store= {store}>
                    <RegisterScreen />
                </Provider>
            </MemoryRouter>
        );

        const { container } = wrapper;

        userEvent.click( container.querySelector('.btn-block') );
        expect( container.querySelector('.auth__alert-error') ).toBeInTheDocument();

    })    

})