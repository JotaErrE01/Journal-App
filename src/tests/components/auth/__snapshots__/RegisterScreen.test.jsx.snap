// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <LoginScreen /> Debe de hacer match con el sanpshot 1`] = `
<div>
  <div
    class="animate__animated animate__fadeIn animate__faster"
  >
    <h3
      class="auth__title mb-5"
    >
      Register
    </h3>
    <form>
      <input
        autocomplete="off"
        class="auth__input mb-5"
        name="name"
        placeholder="Name"
        type="text"
        value=""
      />
      <input
        autocomplete="off"
        class="auth__input mb-5"
        name="email"
        placeholder="Email"
        type="text"
        value=""
      />
      <input
        autocomplete="off"
        class="auth__input mb-5"
        name="password"
        placeholder="Password"
        type="password"
        value=""
      />
      <input
        autocomplete="off"
        class="auth__input mb-5"
        name="password2"
        placeholder="Conifrm Password"
        type="password"
        value=""
      />
      <button
        class="btn btn-primary pointer btn-block mb-5"
        type="submit"
      >
        Register
      </button>
      <a
        class="link"
        href="/auth/login"
      >
        Already Register
      </a>
    </form>
  </div>
</div>
`;
